src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/Delegation/DelegatedCompletionListProvider.cs

```diff
@@ -1,31 +1,28 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the MIT license. See License.txt in the project root for license information.
 
-using System;
 using System.Collections.Generic;
 using System.Collections.Immutable;
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Razor.LanguageServer.Common;
-using Microsoft.AspNetCore.Razor.LanguageServer.Extensions;
 using Microsoft.AspNetCore.Razor.LanguageServer.Protocol;
 using Microsoft.VisualStudio.LanguageServer.Protocol;
 
 namespace Microsoft.AspNetCore.Razor.LanguageServer.Completion.Delegation;
 
 internal class DelegatedCompletionListProvider
 {
-    private static readonly IReadOnlyList<string> s_razorTriggerCharacters = new[] { "@" };
-    private static readonly IReadOnlyList<string> s_cSharpTriggerCharacters = new[] { " ", "(", "=", "#", ".", "<", "[", "{", "\"", "/", ":", "~" };
-    private static readonly IReadOnlyList<string> s_htmlTriggerCharacters = new[] { ":", "@", "#", ".", "!", "*", ",", "(", "[", "-", "<", "&", "\\", "/", "'", "\"", "=", ":", " ", "`" };
+    private static readonly ImmutableHashSet<string> s_razorTriggerCharacters = new[] { "@" }.ToImmutableHashSet();
```

Comment:
Probably of negligible use/importance since these are static readonly, but why do `new[]{...}.ToImmutableHashSet()` instead of `new HashSet<string>{...}.ToImmutableHashSet()` or `ImmutableHashSet.Create(...)`?

Reply:
`new HashSet<string>(...).ToImmutableHashSet()>`. does double the hashing for no value, so I wouldn't do that. Calling `new[] { ... }.ToImmutableHashSet()` is essentially the same as calling `ImmutableHashSet.Create(new [] { ... })`. Between those two, the difference is mostly stylistic, and I opted to keep the fields all looking similar.

Reply:
> new HashSet<string>(...).ToImmutableHashSet()>. does double the hashing for no value

Does it? I would assume `HashSet.ToImmutableHashSet() `wouldn't cause rehashing. It would be similar to `Array<string>.ToImmutableArray ` right?

If it does then TIL

Reply:
Yes, it does. `new HashSet<string>(...).ToImmutableHashSet()` is much more inefficient because it does the work of inserting all of the items into _two_ hash sets, instead of one. Moreover, the `HashSet<T>` constructor takes an `IEnumerable<T>`, so you have to create another collection (probably an array) to fill the `HashSet<T>`. So, you end up creating three collections in all and doing double work.

1. Create a `T[]` containing all items to go into the hash set.
1. Call a `new HashSet<T>(IEnumerable<T>)` which adds all of the items in the `T[]` to the hash set, requiring each item to be hashed into a bucket.
2. Call `ToImmutableHashSet()` which treats the `HashSet<T>` as an `IEnumerable<T>` and adds all of the items from the `HashSet<T>` to an `ImmutableHashSet<T>`. This requires each item to be hashed into buckets again.

Reply:
https://media.tenor.com/ERa47gPVan8AAAAC/the-more-you-know.gif

Reply:
I just looked it up and you're right... I'm very surprised there's no optimization there. There are for cases where you call `ToImmutableHashSet` on an `ImmutableHashSet`... 

😢

Reply:
There's not really much opportunity to optimize, since the internal structure of `HashSet<T>` is pretty different than `ImmutableHashSet<T>`. Also, there's zero opportunity to optimize on .NET Framework 4.7.2, which is most our code runs on in VS. (Remember that `System.Collections.Immutable` came 10+ years after `System.Collections.Generic` was introduced in .NET Framework 2.0.)